Week4/app/src/main/java/com/kulloveth/moviesapp/movies/MovieAdapter.kt

```diff
@@ -0,0 +1,151 @@
+package com.kulloveth.moviesapp.movies
+
+import android.content.Context
+import android.graphics.BitmapFactory
+import android.graphics.Color
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.core.content.ContextCompat
+import androidx.palette.graphics.Palette
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import androidx.viewbinding.ViewBinding
+import com.kulloveth.moviesapp.R
+import com.kulloveth.moviesapp.databinding.HeaderLayoutBinding
+import com.kulloveth.moviesapp.databinding.MovieListItemsBinding
+import com.kulloveth.moviesapp.models.CompositeItem
+import com.kulloveth.moviesapp.models.Movie
+
+
+class MovieAdapter(val movieItemCLickedListener: MovieItemCLickedListener) :
+    ListAdapter<CompositeItem, MovieAdapter.ViewHolder>(MovieDiffCallBack()) {
+
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val movieBinding =
+            MovieListItemsBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        val headBinding =
+            HeaderLayoutBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        return when (viewType) {
+            ViewType.HEADER.ordinal -> ViewHolder(headBinding)
+            ViewType.MOVIE.ordinal -> ViewHolder(movieBinding)
+            else -> throw IllegalArgumentException()
+        }
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(getItem(position))
+        //setup click listener passing the movie through an interface
+        holder.itemView.setOnClickListener {
+            getItem(position).movie?.let { it1 -> movieItemCLickedListener.movieItemCLicked(it1) }
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return if (getItem(position).isHeader) {
+            ViewType.HEADER.ordinal
+        } else {
+            ViewType.MOVIE.ordinal
+        }
+    }
+
+    /**
+     * this a diffutil class which uses your Object type
+     * [CompositeItem] in this class to automatically
+     * manage changes in you adapter without
+     * the need to call notifyDataSetChanged
+     * from every method updating the adapter
+     * */
+    class MovieDiffCallBack : DiffUtil.ItemCallback<CompositeItem>() {
```

Comment:
Could these be turned into overloaded operators? This does seem to work fine though, just a theoretical question.